electrode_models/nucleation_single_particle.py

```diff
@@ -0,0 +1,291 @@
+"""
+    single_particle_electrode.py
+
+    Class file for nucleation and growth (HNG) electrode methods
+"""
+
+import cantera as ct
+import numpy as np
+
+class electrode(): 
+    """
+    Create an electrode object representing the nucleation and growth electrode.
+    """
+
+    def __init__(self, input_file, inputs, sep_inputs, counter_inputs,    
+        electrode_name, params, offset):
+        """
+        Initialize the model.
+        """ 
+   
+        # Import relevant Cantera objects.
+        self.gas_obj = ct.Solution(input_file, inputs['gas-phase'])
+        self.elyte_obj = ct.Solution(input_file, inputs['electrolyte-phase'])
+        self.air_elyte_obj = ct.Interface(input_file, inputs['elyte-iphase'], [self.gas_obj, self.elyte_obj])
+        self.bulk_obj = ct.Solution(input_file, inputs['bulk-phase'])
+        self.product_obj = ct.Solution(input_file, inputs['product-phase'])
+        self.surf_obj = ct.Interface(input_file, inputs['surf-iphase'], 
+            [self.product_obj, self.elyte_obj, self.bulk_obj])
+
+        # Anode or cathode? Positive external current delivers positive charge 
+        # to the anode, and removes positive charge from the cathode.
+        self.name = electrode_name
+        if self.name=='anode':
+            self.i_ext_flag = -1
+        elif self.name=='cathode':
+            self.i_ext_flag = 1
+        else:
+            raise ValueError("Electrode must be an anode or a cathode.")
+
+        # Store the species index of the Li ion in the Cantera object for the 
+        # electrolyte phase:
+        self.index_Li = self.elyte_obj.species_index(inputs['mobile-ion'])
+
+        # Electrode thickness and inverse thickness:
+        self.dy = inputs['thickness']
+        self.dyInv = 1/self.dy
+
+        self.eps_solid = inputs['eps_solid']
+        self.eps_elyte = 1 - self.eps_solid
+        self.eps_oxide = 0
+        
+        self.V_part = 4./3. * np.pi * (inputs['r_p']/ 2)**3  # particle volume [m3]
+        self.A_part = 4. * np.pi * (inputs['r_p'] / 2)**2    # particle surface area [m2]
+        self.A_init = self.eps_solid * self.A_part / self.V_part 
+        self.th_oxide= inputs['th_oxide']
+        self.A_oxide = np.pi* inputs['d_part']**2/4.
+        self.V_oxide = 2./3. * np.pi* (inputs['d_part']/2.)**2 * self.th_oxide
+
+        # For some models, the elyte thickness is different from that of the 
+        # electrode, so we specify is separately:
+        self.dy_elyte = self.dy
+
+        # Phase volume fractions
+        self.eps_solid = inputs['eps_solid']
+        self.eps_elyte = 1 - self.eps_solid
```

Comment:
Might be clearer to put an `_init` on this variable name, if you need the variable at all.  Not certain that you do.